### PR TITLE
C Bindings Compatibility Tweaks

### PR DESCRIPTION
**Pull request by TheBlueMatt**
This serves to revert two of the three bindings-only commits we used for 0.0.103 bindings. Sadly, the last one will likely have to remain.

In doing so, we (finally) give up on having the library be completely bindings-build-agnostic and introduce a (non-feature) `c_bindings` cfg option which very slightly tweaks the API. This lets us make minimally-invasive changes that are bindings-specific while still doing the "right" thing for the Rust API. This also seals `scoring::Time` so that users are no longer able to implement their own `Time`.

After this lands I'll open a PR against the 0.0.103-bindings branch to make similar changes there.

### CODE REVIEW

File: lightning/src/routing/scoring.rs [hunk 1]
```diff
@@ -117,13 +146,34 @@ impl<'a, T: 'a + Score> LockableScore<'a> for RefCell<T> {
 	}
 }
 
-impl<S: Score, T: DerefMut<Target=S>> Score for T {
-	fn channel_penalty_msat(&self, short_channel_id: u64, send_amt_msat: u64, channel_capacity_msat: Option<u64>, source: &NodeId, target: &NodeId) -> u64 {
-		self.deref().channel_penalty_msat(short_channel_id, send_amt_msat, channel_capacity_msat, source, target)
+#[cfg(c_bindings)]
+/// A concrete implementation of [`LockableScore`] which supports multi-threading.
+pub struct ScoreHolder<S: Score> { // TODO(jkczyz): Name this better
```
Comment (jkczyz) [inline lightning/src/routing/scoring.rs:151]: Hmm... given our implementation of `Score` is called `Scorer`, maybe simply name this `LockableScorer`.

**Review by jkczyz — APPROVED** [review]

File: lightning/src/routing/scoring.rs (hunk 1 above)
Comment (TheBlueMatt) [reply lightning/src/routing/scoring.rs:151]: Hmm, true, it is still generic over `Score`, though, so is that a misnomer?

Comment (jkczyz) [reply lightning/src/routing/scoring.rs:151]: How about `AsyncLockableScore`?

Comment (TheBlueMatt) [reply lightning/src/routing/scoring.rs:151]: Wait, what is the `Async` in reference to? Maybe `ConcreteLockableScore` or `MultiThreadedLockableScore`?

Comment (jkczyz) [reply lightning/src/routing/scoring.rs:151]: That it implements the locking using a `Mutex`... guess that's really synchronous. :P `MultiThreadedLockableScore` is fine by me.

**Review by jkczyz — APPROVED** [review]

File: lightning/src/util/logger.rs [hunk 2]
```diff
@@ -86,32 +86,45 @@ impl Level {
...
 	pub level: Level,
+	#[cfg(not(c_bindings))]
 	/// The message body.
 	pub args: fmt::Arguments<'a>,
+	#[cfg(c_bindings)]
+	/// The message body.
+	pub args: String,
 	/// The module path of the message.
-	pub module_path: &'a str,
+	pub module_path: &'static str,
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/util/logger.rs:100]: Hmm, not sure what to say? `'static` just means "string is compiled into the binary, the pointer is valid forever", which is what this is. Ultimately it should never have been `&'a`, but that just effects the borrow checker, not runtime.

**Review by valentinewallace — APPROVED** [review]

**Comment by TheBlueMatt** [comment]
Squashed without diff. Only changes since Jeff's ACK were comment, will land after CI:

```diff
$ git diff-tree -U1 71fe4f519 3539f270c
$ git diff-tree -U1 31a857f 3539f270c
diff --git a/lightning/src/routing/scoring.rs b/lightning/src/routing/scoring.rs
index d12b330e8..e478dbcbe 100644
--- a/lightning/src/routing/scoring.rs
+++ b/lightning/src/routing/scoring.rs
@@ -65,3 +65,3 @@ use sync::{Mutex, MutexGuard};
 
-/// We define Score ever-so-slightly differently based on whether we are being build for C bindings
+/// We define Score ever-so-slightly differently based on whether we are being built for C bindings
 /// or not. For users, `LockableScore` must somehow be writeable to disk. For Rust users, this is
@@ -73,5 +73,5 @@ use sync::{Mutex, MutexGuard};
 /// approach of blindly casting that opaque pointer to a concrete type and calling `Writeable` from
-/// there, but downstream bindings users (e.g. Java users) can't even do that. Instead, we really
-/// want `Score` and `LockableScore` to implement `Writeable` directly, which we do here by
-/// defining `Score` differently for `cfg(c_bindings)`.
+/// there, but other languages downstream of the C bindings (e.g. Java) can't even do that.
+/// Instead, we really want `Score` and `LockableScore` to implement `Writeable` directly, which we
+/// do here by defining `Score` differently for `cfg(c_bindings)`.
 macro_rules! define_score { ($($supertrait: path)*) => {
$
```

**Pull request merged**